### PR TITLE
Add form and models for Phone Numbers

### PR DESCRIPTION
**Pull request by kaizoku**
This PR contains a Registration Extension Form which adds Phone Number and Country Code fields to the User Registration Form
 
**JIRA tickets**: Implements OC-3375.

**Screenshots**:

**Testing instructions**:

1. Clone the repo to `devstack/src` directory: `cd devstack/src; git clone https://github.com/open-craft/edx-custom-registration-form.git`
2. Navigate inside vagrant devstack and install `cd /edx/src/edx-custom-registration-form; pip install -e .`
3. Edit lms.env.json:
    1. Find `REGISTRATION_EXTRA_FIELDS`, change `"city": "hidden"` to `"city": "required"`
    2. Ensure that `"country": "required"` is set. This was set by default for me.
    3. Add a new line above `REGISTRATION_EXTRA_FIELDS` containing: `"REGISTRATION_EXTENSION_FORM": "reg_form_ext.forms.PhoneInfoForm",`
    4. Add a new line containing: `"ADDL_INSTALLED_APPS": ["reg_form_ext"],`
    5. Save and exit editor
4. Restart the LMS by ending restarting in paver, or running `touch ~edxapp/edx-platform/manage.py`
5. Run migrations with `paver update_db`
6. Run tests with `paver test_system -s lms -t /edx/src/edx-custom-registration-form`

**Author notes and concerns**:

1. Including the country codes in `choices.py` is perhaps not the best place for them. I suspect we'd be better of sourcing these from a maintained module such as [django-phonenumber-field](https://github.com/stefanfoulis/django-phonenumber-field) or [python-phonenumbers](https://github.com/daviddrysdale/python-phonenumbers)

**Reviewers**
- [ ] (pomegranited)

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
REGISTRATION_EXTRA_FIELDS:
  city: "required"
EDXAPP_EXTRA_REQUIREMENTS:
  - name: edx-custom-registration-form
    version: 1.0
  - name: git+https://github.com/open-craft/edx-custom-registration-form.git
EDXAPP_ENV_EXTRA:
  REGISTRATION_EXTENSION_FORM: "reg_form_ext.forms.PhoneInfoForm"
```

### CODE REVIEW

**Comment by pomegranited** [comment]
@kaizoku Yes, I agree that using one of the available django phone number libraries is appropriate here.  

From a surface glance, I'm inclined towards https://github.com/stefanfoulis/django-phonenumber-field, because it has nice django fields, widgets and locale support.  I don't quite understand why you'd need to [add `phonenumber-field` to `INSTALLED_APPS`](https://github.com/stefanfoulis/django-phonenumber-field#basic-usage), since there are no URLs, middleware, or models added by this repo.   But if we do, we have a mechanism for adding this along with `ADDL_INSTALLED_APPS`.

Changing this would remove the need for your two separate "country code" field, so you'll need to reset your migrations if you go with `django-phonenumber-field`, so if you decide to go with https://github.com/daviddrysdale/python-phonenumbers instead, that's cool too.

File: reg_form_ext/models.py
Comment (pomegranited) [inline reg_form_ext/models.py]: holds "phone number with country code", not "city and country".

File: reg_form_ext/tests.py
Comment (pomegranited) [inline reg_form_ext/tests.py:13]: Interesting.. you didn't have to set `ADDL_INSTALLED_APPS` here as the [upstream code suggested](https://github.com/edx/edx-platform/pull/10413/files#diff-82bda515879c96ab0641bfcbe58631c3R2672). 

I also manually tested that no `ADDL_INSTALLED_APPS` change was required to run this manually in my ginkgo devstack, so I think it's safe to remove that from your PR description as well.

File: reg_form_ext/choices.py
Comment (pomegranited) [inline reg_form_ext/choices.py:1]: Yeah, for the sake of not maintaining this list, I [agree with using one of the django phonenumber libraries](https://github.com/open-craft/edx-custom-registration-form/pull/1#issuecomment-348395273).

File: reg_form_ext/models.py
Comment (pomegranited) [inline reg_form_ext/models.py:23]: [As noted](https://github.com/open-craft/edx-custom-registration-form/pull/1#issuecomment-348395273), if you use [django-phonenumber-field](https://github.com/stefanfoulis/django-phonenumber-field), you can store these in a single field.

File: reg_form_ext/tests.py
Comment (pomegranited) [reply reg_form_ext/tests.py:13]: And thank you for adding tests 😄

**Review by pomegranited — COMMENTED** [review]
@kaizoku Give it another crack with one of the django phonenumber libraries, and I'll test it again.

But this is looking good 😄   Thanks for the complete PR description and tests.

**Comment by kaizoku** [comment]
Hm, unfortunately it seems like we can't use django-phonenumber-field in the form extension without modifying edx-platform.
When [the form is generated](https://github.com/edx/edx-platform/blob/70dc3359a738eb5ca37d7c233cdc4c0302fc0719/openedx/core/djangoapps/user_api/api.py#L231), they [check it's class in a map](https://github.com/edx/edx-platform/blob/70dc3359a738eb5ca37d7c233cdc4c0302fc0719/openedx/core/djangoapps/user_api/helpers.py#L132) to determine what type of `<input>` to use in the HTML. Naturally this only contains builtin types.. So when I use the `PhoneNumberField`, it's not found in the map, and the code subsequently throws the [ImproperlyConfiguredError](https://github.com/edx/edx-platform/blob/70dc3359a738eb5ca37d7c233cdc4c0302fc0719/openedx/core/djangoapps/user_api/api.py#L233)

I looked into generating this from python-phonenumber, there's also some difficulty there. It has a map of [country codes to region codes](https://github.com/daviddrysdale/python-phonenumbers/blob/ba78a4edc023f7242903cc0ebd65d46f042b67a7/python/phonenumbers/data/__init__.py#L86).
I tried using that list with [pycountry](https://pypi.python.org/pypi/pycountry) to get the full country names, but that has it's own issues. Evidently the phonenumber codes use NATO country codes, which don't map exactly onto ISO-3166 country codes. There's some country codes used in python-phonenumber which aren't in the pycountry list:
```
>>> [ pycountry.countries.get(alpha_2=region[0]) for num, region in phonenumberutil.COUNTRY_CODE_TO_REGION_CODE.items() ]
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pycountry/db.py", line 43, in load_if_needed
    return f(self, *args, **kw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pycountry/db.py", line 101, in get
    return self.indices[field][value]
KeyError: 'AC'
```

So unfortunately, at this point it almost seems better to hardcode the choices as they were..
Unless you have any better ideas?

File: reg_form_ext/tests.py
Comment (kaizoku) [reply reg_form_ext/tests.py:13]: Were you able to run the migrations without including the `ADDL_INSTALLED_APPS` line? I wasn't able to run the migrations without it. So I could load the form, but not submit it.

Ah, this works in the tests because I realize now my tests never saved the form, and so don't test the database or models. I've tried adding a couple tests to rectify that.
However, I can't seem to modify INSTALLED_APPS before the test database is created. So the migrations from this module are not included in the test database.

I tried a few different methods:
* Running syncdb from within `setUpClass()`, based on [this snippet](https://djangosnippets.org/snippets/2843/)
* Adding a `reg_form_ext/test_settings.py`, and configuring that as the `DJANGO_SETTINGS_MODULE` in pytest.ini
* Creating `conftest.py` and overriding pytest fixtures from there including `django_db_setup`, and `pytest_configure`. It seems from [this comment](https://github.com/pytest-dev/pytest-django/issues/33#issuecomment-18058652) that this strategy may not be possible.

It works just fine if I add `INSTALLED_APPS.append('reg_form_ext')` to `lms/envs/test.py`, but that doesn't seem like a reasonable part of the test workflow.

**Comment by pomegranited** [comment]
@kaizoku I found this little library which might help: https://pypi.python.org/pypi/phone-iso3166

But since this is getting more complicated than it should be, and since phone number codes simply don't change, I'm ok with leaving it hard-coded.

**Comment by pomegranited** [comment]
@kaizoku Re your tests, I tried a bunch of things too, and nothing worked 😦 

Could you modify your tests to make them mock out the storage part, like the upstream PR did? https://github.com/edx/edx-platform/pull/10413/files#diff-9c15f7dd12df13f452e1c32f8b68a166R1470

File: reg_form_ext/tests.py
Comment (pomegranited) [reply reg_form_ext/tests.py:13]: @kaizoku Ach you're right -- `ADDL_INSTALLED_APPS` has to be set to run the migrations. 

Re your tests, I tried a bunch of things too, and nothing worked 😦 

We could modify your tests to make them mock out the storage part, [like the upstream PR did](https://github.com/edx/edx-platform/pull/10413/files#diff-9c15f7dd12df13f452e1c32f8b68a166R1470), but that's feeling like more trouble than it's worth.

I'm happy if you just add this test to ensure the fields are required:
```diff
diff --git a/reg_form_ext/tests.py b/reg_form_ext/tests.py
index c0944f6..a261374 100644
--- a/reg_form_ext/tests.py
+++ b/reg_form_ext/tests.py
@@ -1,3 +1,4 @@
+import json
 import unittest
 from django.conf import settings
 from django.test import TestCase
@@ -28,3 +29,14 @@ class PhoneFormTest(TestCase):
         form_data = {"country_code": "1", "phone_number": "206-555-5555"}
         form = PhoneInfoForm(data=form_data)
         self.assertTrue(form.is_valid())
+
+    def test_phone_form_submit(self):
+        """Ensure that form properly requires phone and country_code"""
+        form_data = {}
+        response = self.client.post(reverse('user_api_registration'))
+        response_content = json.loads(response.content)
+        self.assertEquals(response_content.get("country_code"),
+                          [dict(user_message="Please select your country code from the list.")])
+        self.assertEquals(response_content.get("phone_number"),
+                          [dict(user_message="Please enter your phone number")])
+
```

**Review by pomegranited — CHANGES_REQUESTED** [review]
I've added [one suggested test](https://github.com/open-craft/edx-custom-registration-form/pull/1/files#r154519198)

Could you also please add a `README.md` file that describes the project, lists the settings changes required to use it, installation and test instructions.

Will approve once these things are done though, because:

- [x] I tested this on my devstack using the PR instructions.  I ensured that a corresponding PhoneInfo entry was created for my newly-registered user.
- [x] I read through the code
- [x] I checked for accessibility issues -- form is still usable.
- [ ] Includes documentation -- will be handled by README.md

File: README.md
Comment (pomegranited) [inline README.md]: nit: should be:
```yaml
EDXAPP_ENV_EXTRA:
  ADDL_INSTALLED_APPS:
    - reg_form_ext

File: reg_form_ext/tests.py
Comment (pomegranited) [inline reg_form_ext/tests.py]: nit: sorry, you can remove this line.

**Review by pomegranited — APPROVED** [review]
@kaizoku Couple of nits, but you're good to **squash and merge** once they're addressed.

- [x] I tested this on my devstack using the PR instructions.  I ensured that a corresponding PhoneInfo entry was created for my newly-registered user.
- [x] I read through the code
- [x] I checked for accessibility issues -- form is still accessible with new fields.
- [x] Includes documentation -- updated README.md

**Comment by kaizoku** [comment]
@pomegranited thanks for the review and good suggestions on this one!

**Pull request merged**
